prog/checksum.go

```diff
@@ -120,38 +121,114 @@ func encodeStruct(arg *Arg, pid int) []byte {
 	return bytes
 }
 
-func calcChecksumIPv4(arg *Arg, pid int) (*Arg, *Arg) {
-	var csumField *Arg
+func findFieldByName(arg *Arg, name string) *Arg {
 	for _, field := range arg.Inner {
-		if _, ok := field.Type.(*sys.CsumType); ok {
-			csumField = field
-			break
+		if field.Type.FieldName() == name {
+			return field
 		}
 	}
+	return nil
+}
+
+func calcChecksumIPv4(arg *Arg, pid int) (*Arg, *Arg) {
+	csumField := findFieldByName(arg, "csum")
 	if csumField == nil {
 		panic(fmt.Sprintf("failed to find csum field in %v", arg.Type.Name()))
 	}
+	if typ, ok := csumField.Type.(*sys.CsumType); !ok {
+		panic(fmt.Sprintf("checksum field has bad type %v, arg: %+v", csumField.Type, csumField))
+	} else if typ.Kind != sys.CsumIPv4 {
+		panic(fmt.Sprintf("checksum field has bad kind %v, arg: %+v", typ.Kind, csumField))
+	}
 	if csumField.Value(pid) != 0 {
 		panic(fmt.Sprintf("checksum field has nonzero value %v, arg: %+v", csumField.Value(pid), csumField))
 	}
-	bytes := encodeStruct(arg, pid)
+	bytes := encodeArg(arg, pid)
 	csum := ipChecksum(bytes)
 	newCsumField := *csumField
 	newCsumField.Val = uintptr(csum)
 	return csumField, &newCsumField
 }
 
+func extractHeaderParamsIPv4(arg *Arg) (*Arg, *Arg, *Arg) {
+	srcAddr := findFieldByName(arg, "src_ip")
+	if srcAddr == nil {
+		panic(fmt.Sprintf("failed to find src_ip field in %v", arg.Type.Name()))
+	}
+	if srcAddr.Size() != 4 {
+		panic(fmt.Sprintf("src_ip field in %v must be 4 bytes", arg.Type.Name()))
+	}
+	dstAddr := findFieldByName(arg, "dst_ip")
+	if dstAddr == nil {
+		panic(fmt.Sprintf("failed to find dst_ip field in %v", arg.Type.Name()))
+	}
+	if dstAddr.Size() != 4 {
+		panic(fmt.Sprintf("dst_ip field in %v must be 4 bytes", arg.Type.Name()))
+	}
+	protocol := findFieldByName(arg, "protocol")
+	if protocol == nil {
+		panic(fmt.Sprintf("failed to find protocol field in %v", arg.Type.Name()))
+	}
+	if protocol.Size() != 1 {
+		panic(fmt.Sprintf("protocol field in %v must be 1 byte", arg.Type.Name()))
+	}
+	return srcAddr, dstAddr, protocol
+}
+
+func calcChecksumTCP(tcpPacket, srcAddr, dstAddr, protocol *Arg, pid int) (*Arg, *Arg) {
+	tcpHeaderField := findFieldByName(tcpPacket, "header")
+	if tcpHeaderField == nil {
+		panic(fmt.Sprintf("failed to find header field in %v", tcpPacket.Type.Name()))
+	}
+	csumField := findFieldByName(tcpHeaderField, "csum")
+	if csumField == nil {
+		panic(fmt.Sprintf("failed to find csum field in %v", tcpHeaderField.Type.Name()))
+	}
+	if typ, ok := csumField.Type.(*sys.CsumType); !ok {
+		panic(fmt.Sprintf("checksum field has bad type %v, arg: %+v", csumField.Type, csumField))
+	} else if typ.Kind != sys.CsumTCP {
+		panic(fmt.Sprintf("checksum field has bad kind %v, arg: %+v", typ.Kind, csumField))
+	}
+
+	var csum IPChecksum
+	csum.Update(encodeArg(srcAddr, pid))
+	csum.Update(encodeArg(dstAddr, pid))
+	csum.Update(append([]byte{0}, encodeArg(protocol, pid)...))
```

Comment:
Why do you do `append([]byte{0}, ...`?
it is used to make a copy of a slice, but the slice returned by `encodeArg` is already not aliased with anything else.